yarn-project/simulator/src/public/side_effect_trace.ts

```diff
@@ -0,0 +1,296 @@
+import { AvmContractInstanceHint, AvmExecutionHints, AvmExternalCallHint, AvmKeyValueHint, AztecAddress, CallContext, ContractStorageRead, ContractStorageUpdateRequest, EthAddress, Gas, L2ToL1Message, LogHash, NoteHash, Nullifier, ReadRequest } from '@aztec/circuits.js';
+import { UnencryptedFunctionL2Logs, UnencryptedL2Log } from '@aztec/circuit-types';
+import { ContractInstanceWithAddress } from '@aztec/types/contracts';
+import { EventSelector } from '@aztec/foundation/abi';
+import { Fr } from '@aztec/foundation/fields';
+
+import {
+  type PublicExecution,
+  type PublicExecutionResult,
+} from './execution.js';
+import { PublicSideEffectTraceInterface } from './side_effect_trace_interface.js';
+import { createSimulationError } from '../common/errors.js';
+import { AvmExecutionEnvironment } from '../avm/avm_execution_environment.js';
+import { AvmContractCallResults } from '../avm/avm_message_call_result.js';
+
+export type TracedContractInstance = { exists: boolean } & ContractInstanceWithAddress;
+
+export class AvmSideEffectTrace implements PublicSideEffectTraceInterface {
+  /** The side effect counter increments with every call to the trace. */
+  private sideEffectCounter: number; // kept as number until finalized for efficiency
+
+  private contractStorageReads: ContractStorageRead[] = [];
+  private contractStorageUpdateRequests: ContractStorageUpdateRequest[] = [];
+
+  private noteHashReadRequests: ReadRequest[] = [];
+  private newNoteHashes: NoteHash[] = [];
+
+  private nullifierReadRequests: ReadRequest[] = [];
+  private nullifierNonExistentReadRequests: ReadRequest[] = [];
+  private newNullifiers: Nullifier[] = [];
+
+  private l1ToL2MsgReadRequests: ReadRequest[] = [];
+  private newL2ToL1Messages: L2ToL1Message[] = [];
+
+  private unencryptedLogs: UnencryptedL2Log[] = [];
+  private allUnencryptedLogs: UnencryptedL2Log[] = [];
+  private unencryptedLogsHashes: LogHash[] = [];
+
+  private gotContractInstances: ContractInstanceWithAddress[] = [];
+
+  private nestedExecutions: PublicExecutionResult[] = [];
+
+  private avmCircuitHints: AvmExecutionHints;
```

Comment:
Consider making gathering this in a separate tracer (and passing a list of tracers to the sim). It might be useful for tests where you might want a mock tracer for side effects but want the real hints.

In any case, optional, the above case might not exist.

Reply:
I'll give this some thought. That might be nice....

Reply:
Just realized that this might not be possible if you need to share the side effect counter between the tracers (unless you track it in both which is error prone)